kube-runtime/src/coordination.rs

```diff
@@ -0,0 +1,603 @@
+            State::Standby if self.had_error_on_last_try => {
+                self.had_error_on_last_try = false;
+                let rand_val: f64 = rand::thread_rng().gen_range(0.5..1.5);
+                let jitter = rand_val * JITTER_FACTOR * self.config.retry_period.as_secs_f64();
+                return Duration::from_secs_f64(jitter);
+            }
+            // If never observed, or recently delted, then we need to renew now.
+            State::Standby => return Duration::from_secs(0),
+        };
+
+        // Establish how long we need to wait, and return the duration.
+        let now = Instant::now();
+        let deadline = *last_updated + delay;
+        if deadline > now {
+            // Deadline is in the future, so delay until deadline.
+            deadline - now
+        } else {
+            // Else, time to renew now.
+            Duration::from_secs(0)
+        }
+    }
+
+    /// Check if the lease is expired.
+    ///
+    /// If the lease is unknown due to being in state `Standby`, this function will return `true`.
+    fn is_lease_expired(&self) -> bool {
+        match &self.state {
+            State::Leading { last_updated, .. } | State::Following { last_updated, .. } => {
+                let ttl = *last_updated + self.config.lease_duration;
+                ttl <= Instant::now()
+            }
+            State::Standby => true,
+        }
+    }
+}
+
+/// The private state of the leader elector task.
+#[derive(Clone, Debug, PartialEq)]
+enum State {
+    /// This client instance is the leader.
+    Leading {
+        /// The last observed lease state.
+        lease: Lease,
+        /// The last time this lease state updated as leader.
+        last_updated: Instant,
+    },
+    /// A state indicating that a different client is currently the leader, identified by the
+    /// encapsulated string.
+    ///
+    /// When a new leader is detected, this value will be updated with the leader's identity.
+    Following {
+        /// The ID of the current leader.
+        leader: String,
+        /// The last observed lease state.
+        lease: Lease,
+        /// The last time this lease state was observed.
+        last_updated: Instant,
+    },
+    /// A state indicating that the lease state is unknown, does not exist, or that the
+    /// corresponding leader elector task is starting or stopping.
+    Standby,
+}
+
+impl State {
+    /// Get a reference to the last known lease state.
+    fn get_lease(&self) -> Option<&Lease> {
+        match self {
+            Self::Leading { lease, .. } | Self::Following { lease, .. } => Some(lease),
+            Self::Standby => None,
+        }
+    }
+}
+
+/// Different states which a leader elector may be in.
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum LeaderState {
+    /// This client instance is the leader.
+    Leading,
+    /// A state indicating that a different client is currently the leader.
+    Following,
```

Comment:
Should we include the name of the other client in this enum?

Reply:
We could do that, and IIRC an earlier version of this code (before I opened this PR) was indeed doing that. However, I don't think that info is useful at all. Internal to the LeaderElector task, that info is useful. To consumers of the LeaderState stream, it is not useful information for control flow.

I suspect that if folks actually have a use case for tracking this info on the Lease object, then they should probably be watching that least object.

Thoughts?

Reply:
Ok if I mark this as resolved?